map_test.go

```diff
@@ -2,128 +2,513 @@ package swiss
+func TestInitialCapacity(t *testing.T) {
+	testCases := []struct {
+		initialCapacity  int
+		expectedCapacity int
+	}{
+		{0, 0},
+		{1, 7},
+		{7, 7},
+		{8, 15},
+		{1000, 1023},
+		{1024, 2047},
+	}
+	for _, c := range testCases {
+		t.Run("", func(t *testing.T) {
+			m := New[int, int](c.initialCapacity)
+			require.EqualValues(t, c.expectedCapacity, m.capacity)
+		})
+	}
+}
+
+func TestBasic(t *testing.T) {
+	goodHash := func() *Map[int, int] {
+		return New[int, int](0)
+	}
+	badHash := func() *Map[int, int] {
+		m := goodHash()
+		m.hash = func(key unsafe.Pointer, seed uintptr) uintptr {
+			return 0
+		}
+		return m
+	}
+
+	for _, m := range []*Map[int, int]{goodHash(), badHash()} {
+		t.Run("", func(t *testing.T) {
+			e := make(map[int]int)
+			require.EqualValues(t, 0, m.Len())
+			require.EqualValues(t, 0, m.growthLeft)
+
+			// Non-existent.
+			for i := 0; i < 10; i++ {
+				_, ok := m.Get(i)
+				require.False(t, ok)
+			}
+
+			// Insert.
+			for i := 0; i < 10; i++ {
+				m.Put(i, i+10)
+				e[i] = i + 10
+				v, ok := m.Get(i)
+				require.True(t, ok)
+				require.EqualValues(t, i+10, v)
+				require.EqualValues(t, i+1, m.Len())
+				require.Equal(t, e, m.toBuiltinMap())
+			}
+
+			// Update.
+			for i := 0; i < 10; i++ {
+				m.Put(i, i+20)
+				e[i] = i + 20
+				v, ok := m.Get(i)
+				require.True(t, ok)
+				require.EqualValues(t, i+20, v)
+				require.EqualValues(t, 10, m.Len())
+				require.Equal(t, e, m.toBuiltinMap())
+			}
+
+			// Delete.
+			for i := 0; i < 10; i++ {
+				m.Delete(i)
+				delete(e, i)
+				require.EqualValues(t, 10-i-1, m.Len())
+				_, ok := m.Get(i)
+				require.False(t, ok)
+				require.Equal(t, e, m.toBuiltinMap())
+			}
+		})
+	}
+}
+
+func TestRandom(t *testing.T) {
+	goodHash := func() *Map[int, int] {
+		return New[int, int](0)
+	}
+	badHash := func() *Map[int, int] {
+		m := goodHash()
+		m.hash = func(key unsafe.Pointer, seed uintptr) uintptr {
+			return 0
+		}
+		return m
 	}
 
-	b.ResetTimer()
-	var ok bool
-	for i := 0; i < b.N; i++ {
-		_, ok = m.Get(keys[uint32(i)&mod])
+	for _, m := range []*Map[int, int]{goodHash(), badHash()} {
+		t.Run("", func(t *testing.T) {
+			e := make(map[int]int)
+			for i := 0; i < 10000; i++ {
+				switch r := rand.Float64(); {
+				case r < 0.5: // 50% inserts
+					k, v := rand.Int(), rand.Int()
+					if debug {
+						fmt.Printf("insert %d: %d\n", k, v)
+					}
+					m.Put(k, v)
+					e[k] = v
+				case r < 0.65: // 15% updates
+					if k, _, ok := m.randElement(); !ok {
+						require.EqualValues(t, 0, m.Len(), e)
+					} else {
+						v := rand.Int()
+						if debug {
+							fmt.Printf("update %d: %d\n", k, v)
+						}
+						m.Put(k, v)
+						e[k] = v
+					}
+				case r < 0.80: // 15% deletes
+					if k, _, ok := m.randElement(); !ok {
+						require.EqualValues(t, 0, m.Len(), e)
+					} else {
+						if debug {
+							fmt.Printf("delete %d\n", k)
+						}
+						m.Delete(k)
+						delete(e, k)
+					}
+				case r < 0.95: // 25% lookups
```

Comment:
If we have some bug where we somehow drop an element, we won't catch it because `randElement` would never return it.

Since a lookup doesn't change the map, we should look up *all* elements in `e` and make sure they are found. We can do that after each operation that changes the map.

Reply:
Ack. I was thinking about something similar last night. I'm going to add a `Map.verify` method and support for an `invariants` build tag similar to what we do in Pebble. I'm going to do this in a separate PR.